File: src/main/java/pl/training/patterns/behavioral/templatemethod/Application.java
package pl.training.patterns.behavioral.templatemethod;

import java.util.List;

public class Application {

    static class UpperCaseProcessor extends Processor<List<String>, String> {

        @Override
        protected List<String> read() {
            return List.of("template", "method");
        }

        @Override
        protected String process(List<String> data) {
            return String.join(" ", data).toUpperCase();
        }

        @Override
        protected void write(String data) {
            System.out.println(data);
        }

    }

    public static void main(String[] args) {
        new UpperCaseProcessor().run();
    }

}

File: csharp/src/Training.Patterns/Behavioral/TemplateMethod/Application.cs
namespace Training.Patterns.Behavioral.TemplateMethod;

public class Application
{
    internal class UpperCaseProcessor : Processor<IReadOnlyList<string>, string>
    {
        protected override IReadOnlyList<string> Read() => ["template", "method"];

        protected override string Process(IReadOnlyList<string> data) => string.Join(" ", data).ToUpperInvariant();

        protected override void Write(string data)
        {
            Console.WriteLine(data);
        }
    }

    public static void Run()
    {
        new UpperCaseProcessor().Run();
    }
}
